### PR TITLE
1.8.4

### PR DESCRIPTION
**Pull request by lieselwd**
Implements auto acknowledgment on a per-OCA basis with pilots manually selecting which target OCA to 'send' the RCL too. Also implements refactor of DatalinkAuthority. 

**Must run migrations and DatalinkAuthoritiesSeeder upon deployment**

### CODE REVIEW

File: resources/views/controllers/clx/rcl-messages/show.blade.php [hunk 1]
```diff
@@ -1,4 +1,4 @@
-@extends('_layouts.main')
+    @extends('_layouts.main')
```
Comment (williammck) [inline resources/views/controllers/clx/rcl-messages/show.blade.php:1]: nit: where'd this indent come from

File: database/migrations/2024_10_23_105816_implement_dl_authority_foreign_keys.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+        Schema::table('cpdlc_messages', function (Blueprint $table) {
+            $table->string('datalink_authority_id')->default(null)->change();
+            $table->foreign('datalink_authority_id')->references('id')->on('datalink_authorities');
+        });
+    }
+
+    public function down(): void
+    {
+        Schema::table('clx_messages', function (Blueprint $table) {
```
Comment (williammck) [inline database/migrations/2024_10_23_105816_implement_dl_authority_foreign_keys.php:23]: unnecessary table update, pull what's in here out

File: database/migrations/2024_10_23_104818_implement_database_dl_authorities_on_models.php [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        });
+
+        Schema::table('cpdlc_messages', function (Blueprint $table) {
+            $table->renameColumn('datalink_authority', 'datalink_authority_id');
+        });
+    }
+
+    public function down(): void
+    {
+        Schema::table('clx_messages', function (Blueprint $table) {
```
Comment (williammck) [inline database/migrations/2024_10_23_104818_implement_database_dl_authorities_on_models.php:21]: unnecessary table update, pull what's in here out

File: app/Models/DatalinkAuthority.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ * @method static \Illuminate\Database\Eloquent\Builder|DatalinkAuthority whereCreatedAt($value)
+ * @method static \Illuminate\Database\Eloquent\Builder|DatalinkAuthority whereId($value)
+ * @method static \Illuminate\Database\Eloquent\Builder|DatalinkAuthority whereName($value)
+ * @method static \Illuminate\Database\Eloquent\Builder|DatalinkAuthority wherePrefix($value)
+ * @method static \Illuminate\Database\Eloquent\Builder|DatalinkAuthority whereSystem($value)
+ * @method static \Illuminate\Database\Eloquent\Builder|DatalinkAuthority whereUpdatedAt($value)
+ * @method static \Illuminate\Database\Eloquent\Builder|DatalinkAuthority whereValidRclTarget($value)
+ * @mixin \Eloquent
+ */
+class DatalinkAuthority extends Model
```
Comment (williammck) [inline app/Models/DatalinkAuthority.php:31]: where's the migration for this model?

File: app/Http/Controllers/ClxMessagesController.php [hunk 5]
```diff
@@ -160,7 +161,7 @@ public function transmit(RclMessage $rclMessage, ClxMessageRequest $request)
             'entry_time_restriction' => $entryRequirement ?? null,
             'raw_entry_time_restriction' => $request->get('entry_time_requirement'),
             'free_text' => $isReclearance ? '** RECLEARANCE '.now()->format('Hi').' ** '.$request->get('free_text') : $request->get('free_text'),
-            'datalink_authority' => DatalinkAuthorities::from($request->get('datalink_authority')),
+            'datalink_authority_id' => DatalinkAuthority::whereId($request->get('datalink_authority'))->first()->id,
```
Comment (williammck) [inline app/Http/Controllers/ClxMessagesController.php:164]: this seems roundabout and unnecessary? the ID is right there!

File: app/Http/Controllers/ClxMessagesController.php [hunk 6]
```diff
@@ -99,9 +100,9 @@ public function showRclMessage(RclMessage $rclMessage)
 
         return view('controllers.clx.rcl-messages.show', [
             'message' => $rclMessage,
-            'dlAuthorities' => DatalinkAuthorities::cases(),
+            'dlAuthorities' => DatalinkAuthority::whereSystem(false)->get(),
             'tracks' => $rclMessage->is_concorde ? Track::concorde()->get() : Track::active()->get(),
-            'activeDlAuthority' => $this->dataService->getActiveControllerAuthority(Auth::user()) ?? DatalinkAuthorities::NAT,
+            'activeDlAuthority' => $this->dataService->getActiveControllerAuthority(Auth::user()) ?? DatalinkAuthority::whereId('NAT')->first(),
```
Comment (williammck) [inline app/Http/Controllers/ClxMessagesController.php:105]:
```suggestion
            'activeDlAuthority' => $this->dataService->getActiveControllerAuthority(Auth::user()) ?? DatalinkAuthority::find('NAT'),
```
nit

File: app/Http/Controllers/ClxMessagesController.php [hunk 7]
```diff
@@ -280,7 +281,7 @@ public function deleteRclMessage(Request $request, RclMessage $rclMessage)
 
     public function revertToVoice(Request $request, RclMessage $rclMessage)
     {
-        $datalinkAuthority = $this->dataService->getActiveControllerAuthority ?? DatalinkAuthorities::NAT;
+        $datalinkAuthority = $this->dataService->getActiveControllerAuthority ?? DatalinkAuthority::whereId('NAT')->first();
```
Comment (williammck) [inline app/Http/Controllers/ClxMessagesController.php:284]:
```suggestion
        $datalinkAuthority = $this->dataService->getActiveControllerAuthority ?? DatalinkAuthority::find('NAT');
```
nit

File: app/Http/Controllers/ClxMessagesController.php [hunk 8]
```diff
@@ -296,14 +297,14 @@ public function revertToVoice(Request $request, RclMessage $rclMessage)
...
     {
-        $datalinkAuthority = $this->dataService->getActiveControllerAuthority ?? DatalinkAuthorities::NAT;
-        $this->cpdlcService->sendMessage(
-            author: $datalinkAuthority,
-            recipient: $rclMessage->callsign,
-            recipientAccount: $rclMessage->vatsimAccount,
-            message: sprintf(RclResponsesEnum::AcknowledgeMoved->value, strtoupper($datalinkAuthority->description())),
-            caption: RclResponsesEnum::Acknowledge->text()
-        );
+        $datalinkAuthority = $this->dataService->getActiveControllerAuthority ?? DatalinkAuthority::whereId('NAT')->first();
```
Comment (williammck) [inline app/Http/Controllers/ClxMessagesController.php:300]:
```suggestion
        $datalinkAuthority = $this->dataService->getActiveControllerAuthority ?? DatalinkAuthority::find('NAT');
```
nit

File: app/Http/Controllers/ClxMessagesController.php [hunk 9]
```diff
@@ -361,7 +362,7 @@ public function create()
         return view('controllers.clx.rcl-messages.create', [
             'dlAuthorities' => DatalinkAuthorities::cases(),
             'tracks' => Track::where('active', true)->orWhere('concorde', true)->get(),
-            'activeDlAuthority' => $this->dataService->getActiveControllerAuthority(Auth::user()) ?? DatalinkAuthorities::NAT,
+            'activeDlAuthority' => $this->dataService->getActiveControllerAuthority(Auth::user()) ?? DatalinkAuthority::whereId('NAT')->first(),
```
Comment (williammck) [inline app/Http/Controllers/ClxMessagesController.php:365]:
```suggestion
            'activeDlAuthority' => $this->dataService->getActiveControllerAuthority(Auth::user()) ?? DatalinkAuthority::find('NAT'),
```
ditto

File: app/Http/Controllers/ClxMessagesController.php [hunk 10]
```diff
@@ -99,9 +100,9 @@ public function showRclMessage(RclMessage $rclMessage)
 
         return view('controllers.clx.rcl-messages.show', [
             'message' => $rclMessage,
-            'dlAuthorities' => DatalinkAuthorities::cases(),
+            'dlAuthorities' => DatalinkAuthority::whereSystem(false)->get(),
```
Comment (williammck) [inline app/Http/Controllers/ClxMessagesController.php:103]: nit: would this sort of thing be better as a query scope?

File: app/Http/Livewire/Controllers/CreateManualClx.php [hunk 11]
```diff
@@ -36,6 +37,8 @@ class CreateManualClx extends Component
 
     public function mount()
     {
+        $dataService = new VatsimDataService();
+        $this->activeDlAuthority = $dataService->getActiveControllerAuthority(Auth::user()) ?? DatalinkAuthority::whereId('NAT')->first();
```
Comment (williammck) [inline app/Http/Livewire/Controllers/CreateManualClx.php:41]:
```suggestion
        $this->activeDlAuthority = $dataService->getActiveControllerAuthority(Auth::user()) ?? DatalinkAuthority::find('NAT');
```
nit

File: app/Http/Requests/RclMessageRequest.php [hunk 12]
```diff
@@ -88,8 +91,8 @@ public function withValidator($validator)
                 /** Entry fix time requirement */
                 if (config('app.rcl_time_constraints_enabled') && strlen($this->entry_time) == 4) {
                     if (!$this->entryTimeWithinRange($this->entry_time)) {
-                        if (config('app.rcl_auto_acknowledgement_enabled')) {
-                            $this->cpdlcService->sendMessage(author: DatalinkAuthorities::SYS, recipient: $this->callsign, recipientAccount: Auth::user(), message: sprintf(RclResponsesEnum::Contact->value, strtoupper(DatalinkAuthorities::OCEN->description())), caption: RclResponsesEnum::Contact->text());
+                        if (config('app.rcl_auto_acknowledgement_enabled') && DatalinkAuthority::whereId($this->target_datalink_authority_id)->first()->auto_acknowledge_participant) {
```
Comment (williammck) [inline app/Http/Requests/RclMessageRequest.php:94]:
```suggestion
                        if (config('app.rcl_auto_acknowledgement_enabled') && DatalinkAuthority::find($this->target_datalink_authority_id)->auto_acknowledge_participant) {
```
nit

File: app/Http/Requests/RclMessageRequest.php [hunk 13]
```diff
@@ -88,8 +91,8 @@ public function withValidator($validator)
...
+                            $this->cpdlcService->sendMessage(author: DatalinkAuthority::whereId('SYST')->first(), recipient: $this->callsign, recipientAccount: Auth::user(), message: sprintf(RclResponsesEnum::Contact->value, strtoupper(DatalinkAuthorities::OCEN->description())), caption: RclResponsesEnum::Contact->text());
```
Comment (williammck) [inline app/Http/Requests/RclMessageRequest.php:95]:
```suggestion
                            $this->cpdlcService->sendMessage(author: DatalinkAuthority::find('SYST'), recipient: $this->callsign, recipientAccount: Auth::user(), message: sprintf(RclResponsesEnum::Contact->value, strtoupper(DatalinkAuthorities::OCEN->description())), caption: RclResponsesEnum::Contact->text());
```
nit

**Review by williammck — CHANGES_REQUESTED** [review]
sorry, had a bunch of nits and then a couple actual comments

File: app/Models/DatalinkAuthority.php (hunk 4 above)
Comment (lieselwd) [reply app/Models/DatalinkAuthority.php:31]: 450aaba55f6070838d9443f5041298144c9d23ce

File: app/Http/Controllers/ClxMessagesController.php (hunk 5 above)
Comment (lieselwd) [reply app/Http/Controllers/ClxMessagesController.php]: https://github.com/vatsimnetwork/nattrak/pull/160/commits/a662997a44373e51b4d8d90d1f67da610efcdcd2

File: database/migrations/2024_10_23_104818_implement_database_dl_authorities_on_models.php (hunk 3 above)
Comment (lieselwd) [reply database/migrations/2024_10_23_104818_implement_database_dl_authorities_on_models.php:21]: https://github.com/vatsimnetwork/nattrak/pull/160/commits/a662997a44373e51b4d8d90d1f67da610efcdcd2

File: database/migrations/2024_10_23_105816_implement_dl_authority_foreign_keys.php (hunk 2 above)
Comment (lieselwd) [reply database/migrations/2024_10_23_105816_implement_dl_authority_foreign_keys.php:23]: https://github.com/vatsimnetwork/nattrak/pull/160/commits/a662997a44373e51b4d8d90d1f67da610efcdcd2

**Review by williammck — APPROVED** [review]

**Pull request merged**